File: cpp/src/IceUtil/Base64.cpp
// **********************************************************************
//
// Copyright (c) 2001
// MutableRealms, Inc.
// Huntsville, AL, USA
//
// All Rights Reserved
//
// **********************************************************************

#include <IceUtil/Base64.h>
#include <iostream>

using namespace std;

string
IceUtil::Base64::encode(const ByteSeq& plainSeq)
{
    string retval;

    if (plainSeq.size() == 0) 
    {
        return retval;
    }
 
    // Reserve enough space for the returned base64 string
    long base64Bytes = (((plainSeq.size() * 4L) / 3L) + 1L);
    long newlineBytes = (((base64Bytes * 2L) / 76L) + 1L);
    long totalBytes = base64Bytes + newlineBytes;
 
    retval.reserve(totalBytes);

    unsigned char by1 = 0;
    unsigned char by2 = 0;
    unsigned char by3 = 0;
    unsigned char by4 = 0;
    unsigned char by5 = 0;
    unsigned char by6 = 0;
    unsigned char by7 = 0;

    for (unsigned int i = 0; i < plainSeq.size(); i += 3)
    {
        by1 = plainSeq[i];
        by2 = 0;
        by3 = 0;

        if ((i + 1) < plainSeq.size())
        {
            by2 = plainSeq[i+1];
        }

        if ((i + 2) < plainSeq.size())
        {
            by3 = plainSeq[i+2];
        }
 
        by4 = by1 >> 2;
        by5 = ((by1 & 0x3) << 4) | (by2 >> 4);
        by6 = ((by2 & 0xf) << 2) | (by3 >> 6);
        by7 = by3 & 0x3f;

        retval += encode(by4);
        retval += encode(by5);
 
        if ((i + 1) < plainSeq.size())
        {
            retval += encode(by6);
        }
        else
        {
            retval += "=";
        }
 
        if ((i + 2) < plainSeq.size())
        {
            retval += encode(by7);
        }
        else
        {
            retval += "=";
        }
    }

    string outString;
    outString.reserve(totalBytes);
    string::iterator iter = retval.begin();

    while ((retval.end() - iter) > 76)
    {
        copy(iter, iter+76, back_inserter(outString));
        outString += "\r\n";
        iter += 76;
    }

    copy(iter, retval.end(), back_inserter(outString));

    return outString;
}

IceUtil::ByteSeq
IceUtil::Base64::decode(const string& _str)
{
    string str;

    for (unsigned int j = 0; j < _str.length(); j++)
    {
        if (isBase64(_str[j]))
        {
            str += _str[j];
        }
    }

    ByteSeq retval;

    if (str.length() == 0)
    {
        return retval;
    }

    // Figure out how long the final sequence is going to be.
    long lines = _str.size() / 78;
    long totalBytes = (lines * 76) + (((_str.size() - (lines * 78)) * 3) / 4);

    retval.reserve(totalBytes);

    unsigned char by1 = 0;
    unsigned char by2 = 0;
    unsigned char by3 = 0;
    unsigned char by4 = 0;

    char c1, c2, c3, c4;

    for (unsigned int i = 0; i < str.length(); i += 4)
    {
        c1='A';
        c2='A';
        c3='A';
        c4='A';

        c1 = str[i];

        if ((i + 1) < str.length())
        {
            c2 = str[i+1];
        }

        if ((i + 2) < str.length())
        {
            c3 = str[i+2];
        }

        if ((i + 3) < str.length())
        {
            c4 = str[i+3];
        }

        by1 = decode(c1);
        by2 = decode(c2);
        by3 = decode(c3);
        by4 = decode(c4);

        retval.push_back((by1 << 2) | (by2 >> 4));

        if (c3 != '=')
        {
            retval.push_back(((by2 & 0xf) << 4) | (by3 >> 2));
        }

        if (c4 != '=')
        {
            retval.push_back(((by3 & 0x3) << 6) | by4);
        }
    }

    return retval;
}

char
IceUtil::Base64::encode(unsigned char uc)
{
    if (uc < 26)
    {
        return 'A'+uc;
    }
    
    if (uc < 52)
    {
        return 'a'+(uc-26);
    }
    
    if (uc < 62)
    {
        return '0'+(uc-52);
    }
    
    if (uc == 62)
    {
        return '+';
    }

    return '/';
}

unsigned char
IceUtil::Base64::decode(char c)
{
    if (c >= 'A' && c <= 'Z')
    {
        return c - 'A';
    }

    if (c >= 'a' && c <= 'z')
    {
        return c - 'a' + 26;
    }

    if (c >= '0' && c <= '9')
    {
        return c - '0' + 52;
    }

    if (c == '+')
    {
        return 62;
    }
 
    return 63;
}


bool
IceUtil::Base64::isBase64(char c)
{
    if (c >= 'A' && c <= 'Z')
    {
        return true;
    }

    if (c >= 'a' && c <= 'z')
    {
        return true;
    }

    if (c >= '0' && c <= '9')
    {
        return true;
    }

    if (c == '+')
    {
        return true;
    }

    if (c == '/')
    {
        return true;
    }

    if (c == '=')
    {
        return true;
    }

    return false;
}

File: java/src/Ice/src/main/java/com/zeroc/IceUtilInternal/Base64.java
// **********************************************************************
//
// Copyright (c) 2003-2017 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

package com.zeroc.IceUtilInternal;

//
// We would prefer to use java.util.Base64 but unfortunately that class isn't supported in Android until
// Android O, so we are using our own implementation.
//
public class Base64
{
    public static String encode(byte[] plainSeq)
    {
        if(plainSeq == null || plainSeq.length == 0)
        {
            return "";
        }

        int base64Bytes = (((plainSeq.length * 4) / 3) + 1);
        int newlineBytes = (((base64Bytes * 2) / 76) + 1);
        int totalBytes = base64Bytes + newlineBytes;

        StringBuilder retval = new StringBuilder(totalBytes);

        int by1;
        int by2;
        int by3;
        int by4;
        int by5;
        int by6;
        int by7;

        for(int i = 0; i < plainSeq.length; i += 3)
        {
            by1 = plainSeq[i] & 0xff;
            by2 = 0;
            by3 = 0;

            if((i + 1) < plainSeq.length)
            {
                by2 = plainSeq[i+1] & 0xff;
            }

            if((i + 2) < plainSeq.length)
            {
                by3 = plainSeq[i+2] & 0xff;
            }

            by4 = (by1 >> 2) & 0xff;
            by5 = (((by1 & 0x3) << 4) | (by2 >> 4)) & 0xff;
            by6 = (((by2 & 0xf) << 2) | (by3 >> 6)) & 0xff;
            by7 = by3 & 0x3f;

            retval.append(encode((byte)by4));
            retval.append(encode((byte)by5));

            if((i + 1) < plainSeq.length)
            {
                retval.append(encode((byte)by6));
            }
            else
            {
                retval.append('=');
            }

            if((i + 2) < plainSeq.length)
            {
                retval.append(encode((byte)by7));
            }
            else
            {
                retval.append('=');
            }
        }

        StringBuilder outString = new StringBuilder(totalBytes);
        int iter = 0;

        while((retval.length() - iter) > 76)
        {
            outString.append(retval.substring(iter, iter + 76));
            outString.append("\r\n");
            iter += 76;
        }

        outString.append(retval.substring(iter));

        return outString.toString();
    }

    public static byte[] decode(String str)
    {
        StringBuilder newStr = new StringBuilder(str.length());

        for(int j = 0; j < str.length(); j++)
        {
            char c = str.charAt(j);
            if(isBase64(c))
            {
                newStr.append(c);
            }
            else
            {
                throw new IllegalArgumentException("invalid base64 character `" + str.charAt(j) + "' (ordinal " +
                                                   ((int)str.charAt(j)) + ")");
            }
        }

        if(newStr.length() == 0)
        {
            return null;
        }

        // Note: This is how we were previously computing the size of the return
        //       sequence. The method below is more efficient (and correct).
        // size_t lines = str.size() / 78;
        // size_t totalBytes = (lines * 76) + (((str.size() - (lines * 78)) * 3) / 4);

        // Figure out how long the final sequence is going to be.
        int totalBytes = (newStr.length() * 3 / 4) + 1;

        java.nio.ByteBuffer retval = java.nio.ByteBuffer.allocate(totalBytes);

        int by1;
        int by2;
        int by3;
        int by4;

        char c1, c2, c3, c4;

        int pos = 0;
        for(int i = 0; i < newStr.length(); i += 4)
        {
            c1 = 'A';
            c2 = 'A';
            c3 = 'A';
            c4 = 'A';

            c1 = newStr.charAt(i);

            if((i + 1) < newStr.length())
            {
                c2 = newStr.charAt(i + 1);
            }

            if((i + 2) < newStr.length())
            {
                c3 = newStr.charAt(i + 2);
            }

            if((i + 3) < newStr.length())
            {
                c4 = newStr.charAt(i + 3);
            }

            by1 = decode(c1) & 0xff;
            by2 = decode(c2) & 0xff;
            by3 = decode(c3) & 0xff;
            by4 = decode(c4) & 0xff;

            retval.put((byte)((by1 << 2) | (by2 >> 4)));
            ++pos;

            if(c3 != '=')
            {
                retval.put((byte)(((by2 & 0xf) << 4) | (by3 >> 2)));
                ++pos;
            }

            if(c4 != '=')
            {
                retval.put((byte)(((by3 & 0x3) << 6) | by4));
                ++pos;
            }
        }

        byte[] arr = new byte[pos];
        System.arraycopy(retval.array(), 0, arr, 0, pos);
        return arr;
    }

    private static boolean isBase64(char c)
    {
        if(c >= 'A' && c <= 'Z')
        {
            return true;
        }

        if(c >= 'a' && c <= 'z')
        {
            return true;
        }

        if(c >= '0' && c <= '9')
        {
            return true;
        }

        if(c == '+')
        {
            return true;
        }

        if(c == '/')
        {
            return true;
        }

        if(c == '=')
        {
            return true;
        }

        return false;
    }

    private static char encode(byte uc)
    {
        if(uc < 26)
        {
            return (char)('A' + uc);
        }

        if(uc < 52)
        {
            return (char)('a' + (uc - 26));
        }

        if(uc < 62)
        {
            return (char)('0' + (uc - 52));
        }

        if(uc == 62)
        {
            return '+';
        }

        return '/';
    }

    private static byte decode(char c)
    {
        if(c >= 'A' && c <= 'Z')
        {
            return (byte)(c - 'A');
        }

        if(c >= 'a' && c <= 'z')
        {
            return (byte)(c - 'a' + 26);
        }

        if(c >= '0' && c <= '9')
        {
            return (byte)(c - '0' + 52);
        }

        if(c == '+')
        {
            return 62;
        }

        return 63;
    }
}
